browser/scripts/ui-core.js

```diff
@@ -68,6 +97,153 @@ VizorUI.prototype._init = function(e2) {	// called by .init() in ui.js
 };
 
 
+VizorUI.prototype.updateState = function(e) {
+	this.state._update({
+		'chat':		this.dom.chatWindow,
+		'presets':	this.dom.presetsLib,
+		'assets':	this.dom.assetsLib
+	});
+	this.emit(uiEvents.state_updated, this.state._getCopy());
+	return true;
+}
+
+VizorUI.prototype.setupState = function() {
+	var ui = this;
+
+	this.state._update = function(domPanelStates) {	/* @var domPanelStates {key: $domElement, ...} */
+		if (!ui._initialised) return false;
+		// this = ui.state
+		this.context.availWidth = window.screen.availWidth;
+		this.context.availHeight = window.screen.availHeight;
+		var ps = this.panelStates;					// ref
+		if (typeof domPanelStates === 'object') {
+			var key;
+			for (key in domPanelStates) {
+				if (domPanelStates.hasOwnProperty(key)) {
+					ps[key] = VizorUI.getDomPanelState(domPanelStates[key]);
+				}
+			}
+		}
+		return this;
+	};
+
+	this.state._apply = function(newState) {
+		if ((!ui._initialised) || (typeof newState !== 'object')) return false;
+		// this = ui.state
+		this.visible = newState.visible;
+		this.visibility = clone(newState.visibility);
+		ui.applyVisibility();
+		ui.setWorldEditorMode(newState.viewmode === uiViewMode.world_editor);
+
+		if (typeof newState.panelStates === 'undefined') return true;
```

Comment:
same here - when would it be valid to not have a panelStates member in the newState?